src/ledger/LedgerStateCache.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerStateCache.h"
+#include "ledger/LedgerHashUtils.h"
+#include "util/GlobalChecks.h"
+#include "util/Logging.h"
+#include "util/types.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+size_t
+LedgerEntryHash::operator()(LedgerEntry const& entry) const
+{
+    // Extract the key from the entry and return the key's hash.
+    const LedgerKey& key = LedgerEntryKey(entry);
+    return std::hash<stellar::LedgerKey>()(key);
+}
+
+std::optional<LedgerEntry>
+LedgerStateCache::getEntry(LedgerKey const& key) const
+{
+    if (!supportedKeyType(key.type()))
+    {
+        return std::nullopt;
+    }
+
+    std::shared_lock lock(mMutex);
+    if (mState.empty())
+    {
+        CLOG_DEBUG(Ledger, "LedgerStateCache has not yet been populated.");
+        return std::nullopt;
+    }
+
+    auto it = mState.find(KeyToDummyLedgerEntry(key));
+    if (it != mState.end())
+    {
+        return std::optional<LedgerEntry>(*it);
+    }
+    else
+    {
+        // TODO potentially return an optional if we cannot find the entry, but
+        // the cache should contain all entries that are in the ledger so
+        // perhaps a runtime error is appropriate. For now we will just return
+        // nullopt.
+        CLOG_ERROR(Ledger, "LedgerStateCache does not contain entry for key {}",
```

Comment:
This shouldn't be an error. Since we cache everything there are no "misses" such that returning null indicates the entry does not exist.